crates/types/src/string.rs

```diff
@@ -201,16 +180,134 @@ impl StringBuilder {
         debug_assert!(self.inner.len < self.cap);
     }
 
-    /// Build the `String`.
-    #[inline]
-    pub fn finish(self) -> String {
-        let s = String { data: self.inner.data, len: self.inner.len };
+    /// Initialize [`StringBuilder`] with capacity.
+    pub fn with_capacity(cap: usize) -> Self {
+        if cap == 0 {
+            return Self::new();
+        }
+        let real_cap = cap + 1;
+        let ptr = unsafe { libc::malloc(real_cap) };
+        if ptr.is_null() {
+            unable_to_alloc_memory();
+        }
+        Self {
+            inner: String { len: 0, data: ptr as *mut ffi::c_char },
+            cap: real_cap,
+        }
+    }
+
+    /// Reserve space for N more bytes.
+    ///
+    /// Does not allocate if enough space is available.
+    pub fn reserve(&mut self, aditional: usize) {
+        let Some(min_capacity) = self.min_capacity_for_additional(aditional)
+        else {
+            return;
+        };
+        let new_capacity =
+            min_capacity.checked_next_power_of_two().unwrap_or(min_capacity);
+        self.reserve_raw(new_capacity);
+    }
+
+    /// Reserve space for exactly N more bytes.
+    ///
+    /// Does not allocate if enough space is available.
+    pub fn reserve_exact(&mut self, aditional: usize) {
+        if let Some(new_capacity) = self.min_capacity_for_additional(aditional)
+        {
+            self.reserve_raw(new_capacity);
+        }
+    }
+
+    /// Allocate new_capacity bytes.
+    ///
+    /// Accepts [`NonZeroUsize`] allocation size should never be zero for portability reasons.
+    /// A check isn't included as all call sites should already handle the case.
+    fn reserve_raw(&mut self, new_capacity: NonZeroUsize) {
+        let ptr = unsafe {
+            libc::realloc(
+                self.inner.data as *mut ffi::c_void,
+                new_capacity.get(),
+            )
+        };
+        // realloc may return null if it is unable to allocate the requested memory
+        if ptr.is_null() {
+            unable_to_alloc_memory();
+        }
+        self.inner.data = ptr as *mut ffi::c_char;
+        self.cap = new_capacity.get();
+    }
+
+    /// Finish building the [`String`] but do not shrink the allocation
+    ///
+    /// Useful if we already know that the allocation does not have extra capacity.
+    ///
+    /// In debug mode checks if any of the following are true:
+    /// - data ptr is null with a capacity and length of zero.
+    /// - ptr points to an allocation and capacity > length.
+    /// - ptr points to an allocation with a capacity > 1 (if the capacity is 1 we are creating the
+    ///     allocation just to store a null byte, we should return null instead)
```

Comment:
It made sense to include this info when the function was private, but now this is public and callers don't care about any of it.